### PR TITLE
[FLINK] Fix multiple executions lead to abnormal results on Flink 1.14

### PR DESCRIPTION
**Pull request by waywtdcc**
Fix bug in flink 1.14 version, multiple executions lead to abnormal results

### _Why are the changes needed?_

Fix bug in flink 1.14 version, multiple executions lead to abnormal results

### _How was this patch tested?_
- [X] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [X] Add screenshots for manual tests if appropriate

- [X] [Run test](https://kyuubi.readthedocs.io/en/master/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: externals/kyuubi-flink-sql-engine/src/main/java/org/apache/kyuubi/engine/flink/result/ResultSet.java [hunk 1]
```diff
@@ -108,12 +108,32 @@ public String toString() {
...
+        TableResultImpl.builder()
+            .resultKind(ResultKind.SUCCESS)
+            .schema(ResolvedSchema.of(Column.physical("result", DataTypes.STRING())))
+            .data(Collections.singletonList(Row.of("OK")))
+            .build();
+    return result;
+  }
+
   public static ResultSet fromTableResult(TableResult tableResult) {
+    boolean flinkVersionAtLeast = FlinkEngineUtils.isFlinkVersionAtLeast("1.15");
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/main/java/org/apache/kyuubi/engine/flink/result/ResultSet.java:127]: if there is no reuse case, it's better to inline the variables and functions

File: externals/kyuubi-flink-sql-engine/src/main/java/org/apache/kyuubi/engine/flink/result/ResultSet.java [hunk 2]
```diff
@@ -108,12 +108,32 @@ public String toString() {
         + '}';
   }
 
+  /**
+   * Before version 1.15, there were concurrency issues.
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/main/java/org/apache/kyuubi/engine/flink/result/ResultSet.java:112]: Could you please add a JIRA ticket in the comment? e.g.
"Before FLINK-XXXX, there were concurrency issues, which affects Flink 1.14"

**Comment by pan3793** [comment]
@waywtdcc thanks for digging out, left some general comments, and please run `dev/reformat` to quickly fix the style issue. cc @link3280 as he is the Flink expert.

File: externals/kyuubi-flink-sql-engine/src/main/java/org/apache/kyuubi/engine/flink/result/ResultSet.java [hunk 3]
```diff
@@ -108,12 +108,32 @@ public String toString() {
...
+   *
+   * @return
```
Comment (link3280) [inline externals/kyuubi-flink-sql-engine/src/main/java/org/apache/kyuubi/engine/flink/result/ResultSet.java:114]: Please update the javadoc.

File: externals/kyuubi-flink-sql-engine/src/test/scala/org/apache/kyuubi/engine/flink/operation/FlinkOperationSuite.scala [hunk 4]
```diff
@@ -17,27 +17,25 @@
...
 import org.apache.kyuubi.engine.flink.util.TestUserClassLoaderJar
 import org.apache.kyuubi.jdbc.hive.KyuubiStatement
-import org.apache.kyuubi.operation.{HiveJDBCTestHelper, NoneMode}
 import org.apache.kyuubi.operation.meta.ResultSetSchemaConstant._
+import org.apache.kyuubi.operation.{HiveJDBCTestHelper, NoneMode}
 import org.apache.kyuubi.service.ServiceState._
+import org.scalatest.concurrent.PatienceConfiguration.Timeout
+import org.scalatest.time.SpanSugar._
+
+import java.sql.DatabaseMetaData
```
Comment (link3280) [inline externals/kyuubi-flink-sql-engine/src/test/scala/org/apache/kyuubi/engine/flink/operation/FlinkOperationSuite.scala:36]: Please remove unused imports

**Review by link3280 — CHANGES_REQUESTED** [review]
@waywtdcc Hi, thanks for your PR! WRT to `abnormal results`, I guess you mean if we execute multiple SQLs that return `OK`, the second and latter results would be empty. FYI, this should be a [known bug](https://issues.apache.org/jira/browse/FLINK-25558) in Flink 1.14 and was fixed in Flink 1.15. 

It might need a few changes in the code style as @pan3793 mentioned. I left some comments, PTAL.

**Comment by waywtdcc** [comment]
> dev/reformat

Does kyuubi have check style xml? On my side, I passed mvn spotless:apply
verify

**Comment by pan3793** [comment]
The CI reports
```
[ERROR] Failed to execute goal org.scalastyle:scalastyle-maven-plugin:1.0.0:check (default-cli) on project kyuubi-flink-sql-engine_2.12: Failed during scalastyle execution: You have 7 Scalastyle violation(s). -> [Help 1]
```

Unfortunately, scalastyle can not be fixed automatically

File: externals/kyuubi-flink-sql-engine/src/main/java/org/apache/kyuubi/engine/flink/result/ResultSet.java [hunk 5]
```diff
@@ -108,7 +113,26 @@ public String toString() {
         + '}';
   }
 
+  /**
+   * Before version 1.15(FLINK-25558), there were concurrency issues.
+   *
+   * @return TableResult.
+   */
+  public static TableResult createResultOk() {
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/main/java/org/apache/kyuubi/engine/flink/result/ResultSet.java:121]: please inline this function and comments

File: externals/kyuubi-flink-sql-engine/src/main/java/org/apache/kyuubi/engine/flink/result/ResultSet.java [hunk 6]
```diff
@@ -49,6 +53,7 @@ public class ResultSet {
   // list of boolean in streaming mode,
   // true if the corresponding row is an append row, false if its a retract row
   private final List<Boolean> changeFlags;
+  public static boolean versionAtLeast15 = FlinkEngineUtils.isFlinkVersionAtLeast("1.15");
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/main/java/org/apache/kyuubi/engine/flink/result/ResultSet.java:56]: Pullout everything to the field is not a good practice. If logic/variables are not complex and not used multi times, please keep them inline and local scope.

Comment (waywtdcc) [reply externals/kyuubi-flink-sql-engine/src/main/java/org/apache/kyuubi/engine/flink/result/ResultSet.java:56]: Then restore the original and put it in the fromTableResult method?

Comment (pan3793) [reply externals/kyuubi-flink-sql-engine/src/main/java/org/apache/kyuubi/engine/flink/result/ResultSet.java:56]: The following style should be clear enough.
```
static import org.apache.kyuubi.engine.flink.FlinkEngineUtils.isFlinkVersionAtLeast

if (isFlinkVersionAtLeast("1.15")) {
  ...
} else {
  ...
}
```

**Comment by link3280** [comment]
@waywtdcc Please fix the CI issue first. 

This issue looks weird to me. `FlinkEngineUtils` not found. @pan3793 Do you have any clue?

**Comment by pan3793** [comment]
> ... `FlinkEngineUtils` is not found

Maybe caused by mixed-using Java/Scala

**Comment by pan3793** [comment]
To get out of the compile issue cause by Java/Scala mix reference, I migrate the Java code to Scala code, and improved the code by using `Dyn*` reflection tool, the change passed in local test, it suppose to pass the CI cross version verification, let wait for CI results.

**Comment by pan3793** [comment]
CI passed, and there is no logic change since @waywtdcc last commit.

Thanks all, merged to master/1.7

**Comment by pan3793** [comment]
@waywtdcc would you please link you email `chenchao4@grgbanking.com` to GitHub account? Otherwise your contribution will not be counted by GitHub

**Pull request closed without merging**